src/ui/src/components/workflows/WorkflowsWorkflow.vue

```diff
@@ -0,0 +1,283 @@
+const { getComponentInfoFromDrag } = useDragDropComponent(wf);
+
+const activeNodeOut: Ref<{
+	fromComponentId: string;
+	outId: string;
+} | null> = ref(null);
+
+function refreshArrows() {
+	const a = [];
+	const canvasCBR = rootEl.value?.getBoundingClientRect();
+	if (!canvasCBR) {
+		return;
+	}
+	const nodes = wf.getComponents(workflowComponentId);
+
+	nodes
+		.filter((node) => node.outs?.length > 0)
+		.forEach((node) => {
+			const fromNodeId = node.id;
+			node.outs.forEach((out) => {
+				const fromEl = document.querySelector(
+					`[data-writer-id="${fromNodeId}"] [data-writer-socket-id="${out.outId}"]`,
+				);
+				const toEl = document.querySelector(
+					`[data-writer-id="${out.toNodeId}"]`,
+				);
+				if (!fromEl || !toEl) return;
+				const fromCBR = fromEl.getBoundingClientRect();
+				const toCBR = toEl.getBoundingClientRect();
+				a.push({
+					x1: fromCBR.x - canvasCBR.x + fromCBR.width / 2,
+					y1: fromCBR.y - canvasCBR.y + fromCBR.height / 2,
+					x2: toCBR.x - canvasCBR.x,
+					y2: toCBR.y - canvasCBR.y + toCBR.height / 2,
+					color: getComputedStyle(fromEl).backgroundColor,
+				});
+			});
+		});
+	arrows.value = a;
+}
+
+const isUnselectable = computed(() => {
+	if (activeNodeOut.value === null) return null;
+	return true;
+});
+
+function handleNodeOutSelect(componentId: Component["id"], outId: string) {
+	activeNodeOut.value = {
+		fromComponentId: componentId,
+		outId,
+	};
+}
+
+function handleDragstart(ev: DragEvent) {
+	var img = new Image();
+	img.src =
+		"data:image/gif;base64,R0lGODlhAQABAIAAAAUEBAAAACwAAAAAAQABAAACAkQBADs=";
+	ev.dataTransfer.setDragImage(img, 0, 0);
+	clickOffset = getAdjustedCoordinates(ev);
+}
+
+function handleDrag(ev: DragEvent) {
+	ev.preventDefault();
+	const { x, y } = getAdjustedCoordinates(ev);
+	if (x < 0 || y < 0) return;
+	renderOffset.value.x = Math.max(
+		0,
+		renderOffset.value.x - (x - clickOffset.x),
+	);
+	renderOffset.value.y = Math.max(
+		0,
+		renderOffset.value.y - (y - clickOffset.y),
+	);
+	refreshArrows();
+}
+
+function handleDragover(ev: DragEvent) {
+	ev.preventDefault();
+	ev.stopPropagation();
+}
+
+function getAdjustedCoordinates(ev: DragEvent) {
+	const canvasCBR = rootEl.value.getBoundingClientRect();
+	const x = ev.pageX - canvasCBR.x + renderOffset.value.x;
+	const y = ev.pageY - canvasCBR.y + renderOffset.value.y;
+	return { x, y };
+}
+
+function handleDrop(ev: DragEvent) {
+	ev.preventDefault();
+	ev.stopPropagation();
+	const dropInfo = getComponentInfoFromDrag(ev);
+
+	if (!dropInfo) return;
+	const { draggedType, draggedId } = dropInfo;
+	if (draggedId) return;
+
+	const { x, y } = getAdjustedCoordinates(ev);
+	if (x < 0 || y < 0) return;
+
+	createNode(draggedType, x, y);
+}
+
+function handleArrowClick() {}
+
+function handleNodeDragStart(ev: DragEvent) {
+	ev.stopPropagation();
+	var img = new Image();
+	img.src =
+		"data:image/gif;base64,R0lGODlhAQABAIAAAAUEBAAAACwAAAAAAQABAAACAkQBADs=";
+	ev.dataTransfer.setDragImage(img, 0, 0);
```

Comment:
I think you can refactor to reuse 

https://github.com/writer/writer-framework/blob/0e142ab1446820b2b7e8aed75e25a961bd5e7fd0/src/ui/src/components/workflows/WorkflowsWorkflow.vue#L139-L143

Maybe introducing a function like `getDragImage`

Reply:
Implemented